### ISSUE TITLE
How to handle concrete and symbolic byte-array differently?

### ISSUE THREAD

**Issue by yzhang90** [kframework/kore#1371]
We have the following rules
```
    syntax ByteArray ::= ByteArray "[" Int ".." Int "]" [function]
 // --------------------------------------------------------------
    rule WS [ START .. WIDTH ] => substrBytes(padRightBytes(WS, START +Int WIDTH, 0), START, START +Int WIDTH) requires START <Int #sizeByteArray(WS) [concrete]
    rule WS [ START .. WIDTH ] => padRightBytes(.Bytes, WIDTH, 0)   [owise, concrete]
```
for concrete bytes.
If bytes are symbolic, we have the following 2 functions to wrap it:
```
    //Byte array buffer. Lemmas defined in evm-data-symbolic.k
    // SIZE, DATA // left zero padding
    syntax ByteArray ::= #buf ( Int , Int ) [function, functional, smtlib(buf)]
    syntax ByteArray ::= #bufSeg ( ByteArray , Int , Int ) [function, smtlib(bufSeg)] // BUFFER, START, WIDTH
```
and I'd like to introduce the following rule to lift the bytearray accessing operation
```
    rule [lift]: BUF[START .. WIDTH] => #bufSeg(BUF, START, WIDTH)
       requires #range(0 <= START < lengthBytes(BUF))
       andBool #range(0 < WIDTH <= lengthBytes(BUF) -Int START)
```
However, if `BUF` is concrete, it will be conflict with the rules above.

In the javabackend, we introduce the following predicate
```
    syntax Bool ::= #isBuf ( ByteArray ) [function]
    rule #isBuf(#buf(_,_)) => true
    rule #isBuf(#bufSeg(_,_,_)) => true
```
, and say that the `lift` rule can only be applied when `#isBuf(BUF)` is true.
However, this doesn't really make sense logically as `#buf` and `#bufSeg` are not constructors of the ByteArray.

**Comment by yzhang90** [comment]
solution 1:
introduce `non-concrete` attribute, which is opposite of `concrete` attribute and mark `lift` rule as `non-concrete`

solution 2.
duplicate the `lift` rule. replace `BUF` with `#buf(...)` and `#bufSeg(...)` respectively.

solution 3.
introduce `matches` attribute

solution 4.
introduce priorities on functional rules.

**Comment by ttuegel** [comment]
Fixed by: #1389
